[spring-week5] 유효성 검사

goodcodings opened the pull request:
2022-04-20
feat/유저 DTO생성

goodcodings committed:
feat/유저관리 DTO생성 및 패키지 구조 변경

goodcodings commented on app/src/main/java/com/codesoom/assignment/domain/policy/Policy.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+
+public class Policy {
+    private final List<PolicyStrategy> policyStrategies;
```
유효성 검사시 요구사항 변경시 SOLID의 OCP의 깨짐을 최소화 하기 위해 
컬럼(비밀번호,유저이름)마다 전략패턴을 적용하였습니다

goodcodings commented on app/src/main/java/com/codesoom/assignment/dto/user/UserData.java:
```diff
@@ -0,0 +1,56 @@
+    @Getter
+    @Builder
+    @ToString
+    public static class RegisterUserRequest {
```
클라이언트로부터 등록,삭제, 수정에 필요한 컬럼이 다를 수 있어 각각의 객체를 작성했으며 thread-safe하기 위해 중첩 정적 클래스를 작성했습니다

johngrib replied:
이렇게 자신의 코드를 설명하는 것은 좋은 습관입니다. 혼자서 작업할 때에도 수시로 해 보세요.

johngrib reviewed (commented):
PR 대상이 `CodeSoom:main` 으로 되어 있습니다. 수정해 주세요. 최상단 오른쪽에 있는 `Edit` 버튼을 클릭해서 바꿀 수 있습니다.

![image](https://user-images.githubusercontent.com/1855714/164225686-83631db4-ecd0-43d3-9ba7-64a04a94f8c8.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/policy/PasswordPolicy.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.util.Assert;
+
+
+public class PasswordPolicy implements PolicyStrategy {
```
좀 더 나은 클래스 이름도 가능할 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/policy/PasswordPolicy.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    private void checkPasswordEmpty(String email) {
+        Assert.hasText(email, "비밀번호는 공백 허용이 되지 않습니다");
```
`허용이 되지 않습니다`보다 더 나은 표현을 고민해 보세요.

johngrib reviewed (commented):
이제는 커밋의 단위도 고려해보도록 합시다. 지금은 커밋 하나에 담긴 주제가 다양합니다. 여러 주제별로 커밋을 나누어 해 보세요.

![image](https://user-images.githubusercontent.com/1855714/164226432-eb856f14-2872-41a6-b4a1-60bea63b29bf.png)

johngrib reviewed (commented):
유효성검사를 검증하는 테스트 코드가 있어야겠습니다

goodcodings committed:
refactor:클래스명 변경

기존에 유효성 검사하는 클래스명 변경
1. Policy -> ValidatorExecutor
2. PolicyStrategy -> Validator
3. PasswordPolicy -> PasswordValidator

goodcodings committed:
refactor:controller 패키지 내부 도메인 별로 분리

각 도메인 마다 패키지 생성 후 분리

goodcodings committed:
refactor:application layer내 도메인 별로 분리

각 도메인 마다 패키지 생성 후 분리

goodcodings committed:
feature:user domain 기능 정의

goodcodings committed:
refactor:presentation layer내  dto클래스명 변경

goodcodings committed:
feature:dto 유효성검사 코드 작성중

goodcodings commented:
dto의 정의된 annotation만 유효성 검사를 해보고 싶은데... 방법을 찾고있어서 정리되면 공유하겠습니다 ㅠㅠ

goodcodings commented:
과제를 진행하다가 presentation layer의 controller가 application layer의 class를 의존하고 있는데 class를 의존하게 되면 DIP가 잘 지켜지지 않을 수도 있다고 생각이 듭니다...ㅠㅠ
DIP는 제가 알기론 추상화(인터페이스)에 의존해야하며, 구체화(구현체)에 의존하면 안된다는 원칙으로 알고 있는데... 
제가 혹시 오해하고 잘못 이해하고 있는지 궁금합니다..

johngrib commented:
> 과제를 진행하다가 �presentation layer의 controller가 application layer의 class를 의존하고 있는데 class를 의존하게 되면 DIP가 잘 지켜지지 않을 수도 있다고 생각이 듭니다...ㅠㅠ DIP는 제가 알기론 추상화(인터페이스)에 의존해야하며, 구체화(구현체)에 의존하면 안된다는 원칙으로 알고 있는데... 제가 혹시 오해하고 잘못 이해하고 있는지 궁금합니다..

오해도 아니고 잘못 이해한 것도 아닙니다. 실제로 업무에서도 서비스 클래스를 직접 주입받아서 쓰는 것보다 인터페이스를 선언하고 인터페이스의 구현체를 코딩해서 만든 다음, 주입받을 bean으로는 인터페이스를 선언하는 방식을 많이 쓰죠. 물론 구현체가 달랑 하나 뿐이고 오랫동안 다른 구현체가 필요할 일이 없거나 너무 단순한 프로젝트라면 굳이 인터페이스를 선언하지 않기도 하겠죠. 뭐든지 선택이니까요. 어느쪽 방식이 항상 옳은 것은 아니니까요.

그런데 제가 묻고 싶은 게 있습니다. 그렇게 생각하시는데 인터페이스를 사용하지 않고 구체 클래스를 의존하도록 코딩하신 이유는 무엇인가요? 

![image](https://user-images.githubusercontent.com/1855714/164460623-ba862bd9-0948-4d19-b511-38e46b7c5451.png)

아는 대로 작업하지 않은 이유는 무엇인가요? 그리고 DIP를 지켜야 하는 이유는 무엇일까요?

johngrib commented on app/src/main/java/com/codesoom/assignment/application/user/UserService.java:
```diff
@@ -0,0 +1,40 @@
+    public User updateUser(UserData.UpdateUserRequest request){
+        User user = findUser(request.getId());
+
+        user.change(
+                request.getUsername(),
+                request.getEmail(),
+                request.getPassword()
+        );
```
상태를 변경하는 방법을 이번에 사용해본 셈이군요. 이번에는 상태를 변경하지 않고 주어진 값을 갖는 새로운 인스턴스를 리턴하는 방법을 연습해 봅시다. `user.change`가 리턴값을 갖게 하는 거죠.

goodcodings replied:
JPA는 영속성 컨텍스트에서 영속화 되어있는 데이터의 경우 바뀌게 되면 dirty cheching으로 인해 자동으로 반영된다고 공부를 하면서 알게 됬는데 새로운 인스턴스를 리턴하게 되면 이점이 있는지 궁금합니다!

johngrib replied:
JPA를 염두에 둔 제안은 아니에요. JPA가 관리하는 것을 의도하셨다면 그대로 두어도 문제 없죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/user/UserService.java:
```diff
@@ -0,0 +1,40 @@
+
+    private User findUser(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() -> new UserNotFoundException(id));
```
`id`를 입력받게 하지 말고 구체적인 에러 메시지를 입력하도록 바꾼 다음, 에러 메시지를 곰곰히 생각해서 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/common/validator/PasswordValidator.java:
```diff
@@ -0,0 +1,18 @@
+public class PasswordValidator implements ConstraintValidator<Password,String> {
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        ValidatorExecutor validatorExecutor =new ValidatorExecutor(
```
```suggestion
        ValidatorExecutor validatorExecutor = new ValidatorExecutor(
```

johngrib commented on app/src/main/java/com/codesoom/assignment/common/validator/PasswordValidator.java:
```diff
@@ -0,0 +1,18 @@
+        ValidatorExecutor validatorExecutor =new ValidatorExecutor(
+                new com.codesoom.assignment.domain.validator.PasswordValidator(value)
+        );
+        boolean isValid = validatorExecutor.executePolicyStrategy();
+        return isValid;
```
좋은 습관 하나: IDE의 inspector를 실행해보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/user/UserController.java:
```diff
@@ -0,0 +1,23 @@
+import com.codesoom.assignment.dto.user.UserData;
+import lombok.RequiredArgsConstructor;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
```
`import *`을 사용하지 않는 습관을 만들어 봅시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/AbstractEntity.java:
```diff
@@ -0,0 +1,20 @@
+@Getter
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public class AbstractEntity {
```
에릭 에반스의 "도메인 주도 설계"에서 `Entity`가 나오는 모든 페이지를 찾아 읽어보세요. 그리고 이 클래스 이름이 적절한지 고민해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/validator/PasswordValidator.java:
```diff
@@ -0,0 +1,22 @@
+        this.password = password;
+    }
+
+    private void checkPasswordEmpty(String email) {
+        Assert.hasText(email, "비밀번호는 공백 허용이 되지 않습니다");
+    }
```
연속된 숫자도 허용하지 않아야 한다면 어떻게 할 수 있을까요?

eggNnuts commented:
> > 과제를 진행하다가 �presentation layer의 controller가 application layer의 class를 의존하고 있는데 class를 의존하게 되면 DIP가 잘 지켜지지 않을 수도 있다고 생각이 듭니다...ㅠㅠ DIP는 제가 알기론 추상화(인터페이스)에 의존해야하며, 구체화(구현체)에 의존하면 안된다는 원칙으로 알고 있는데... 제가 혹시 오해하고 잘못 이해하고 있는지 궁금합니다..
> 
> 오해도 아니고 잘못 이해한 것도 아닙니다. 실제로 업무에서도 서비스 클래스를 직접 주입받아서 쓰는 것보다 인터페이스를 선언하고 인터페이스의 구현체를 코딩해서 만든 다음, 주입받을 bean으로는 인터페이스를 선언하는 방식을 많이 쓰죠. 물론 구현체가 달랑 하나 뿐이고 오랫동안 다른 구현체가 필요할 일이 없거나 너무 단순한 프로젝트라면 굳이 인터페이스를 선언하지 않기도 하겠죠. 뭐든지 선택이니까요. 어느쪽 방식이 항상 옳은 것은 아니니까요.
> 
> 그런데 제가 묻고 싶은 게 있습니다. 그렇게 생각하시는데 인터페이스를 사용하지 않고 구체 클래스를 의존하도록 코딩하신 이유는 무엇인가요?
> 
> ![image](https://user-images.githubusercontent.com/1855714/164460623-ba862bd9-0948-4d19-b511-38e46b7c5451.png)
> 
> 아는 대로 작업하지 않은 이유는 무엇인가요? 그리고 DIP를 지켜야 하는 이유는 무엇일까요?

!

goodcodings commented:
> > 과제를 진행하다가 �presentation layer의 controller가 application layer의 class를 의존하고 있는데 class를 의존하게 되면 DIP가 잘 지켜지지 않을 수도 있다고 생각이 듭니다...ㅠㅠ DIP는 제가 알기론 추상화(인터페이스)에 의존해야하며, 구체화(구현체)에 의존하면 안된다는 원칙으로 알고 있는데... 제가 혹시 오해하고 잘못 이해하고 있는지 궁금합니다..
> 
> 오해도 아니고 잘못 이해한 것도 아닙니다. 실제로 업무에서도 서비스 클래스를 직접 주입받아서 쓰는 것보다 인터페이스를 선언하고 인터페이스의 구현체를 코딩해서 만든 다음, 주입받을 bean으로는 인터페이스를 선언하는 방식을 많이 쓰죠. 물론 구현체가 달랑 하나 뿐이고 오랫동안 다른 구현체가 필요할 일이 없거나 너무 단순한 프로젝트라면 굳이 인터페이스를 선언하지 않기도 하겠죠. 뭐든지 선택이니까요. 어느쪽 방식이 항상 옳은 것은 아니니까요.
> 
> 그런데 제가 묻고 싶은 게 있습니다. 그렇게 생각하시는데 인터페이스를 사용하지 않고 구체 클래스를 의존하도록 코딩하신 이유는 무엇인가요?
> 
> ![image](https://user-images.githubusercontent.com/1855714/164460623-ba862bd9-0948-4d19-b511-38e46b7c5451.png)
> 
> 아는 대로 작업하지 않은 이유는 무엇인가요? 그리고 DIP를 지켜야 하는 이유는 무엇일까요?

**1. 먼저 아는 대로 작업하지 않은 이유** 
 
제가 생각한 Domain layer의 구성은 `공연, 할인 정책, 예매`등으로 구성된 **도메인 모델**과 `할인 금액 계산하기, 예매하기`등과 객체간의 협력에 필요한 **추상화 레벨이 높은 메세지 목록**이 나열된 인터페이스가 존재 해야하며 

application layer의 구성은 domain layer에서 나열된 **메세지의 목록**을 어떻게 풀어 나가야할지 자세한 구현을 해야하는 layer로 잘못 이해를 하고 있었습니다.

그래서 코딩을 할 당시에 presentation layer의 구성요소중 하나인 컨트롤러에 application layer에 class를 아닌 
domain layer에 인터페이스를 의존해야한다고 생각하게 됬습니다.
그런데  DDD의 의존계층을 보게되면 presentation layer은 application layer을 의존해야하고 application layer은 domain layer을 해야 한다고 되어있고 제가 생각한대로 presentation layer에 domain layer의 인터페이스를 표기하면 의존 관계가 엉망이 된다고 생각하게 되어 많이 혼란스러워서  어떻게 구현해야할지 몰라 기존에 작성한 ProductService 형태를 따라서 코딩을 하게 됬습니다.

**2. DIP를 지켜야하는 이유**

DIP를 지켜야하는 이유는 Loose Coupling(느슨한 결합)을 위해서 지켜야한다고 생각합니다.
추상화 레벨이 낮은 구현체는 언제든지 변할 수 있는 가능성이 농후 하다고 생각합니다.
만약 구현체를 변경하게 되면 그 구현체를 의존하고 있는 모든 객체에  영향을 미치게 되고 어떠한 사이드 이펙트가 발생할지 모른다고 생각합니다.
DIP를 지키지 않는 다는 말은 연극의 배역을 연기하는 배우에게 의존한다고 생각할 수 있을거 같습니다.
만약 배우에게 무슨일이 생기게되면 연극을 할 수 없는 끔직한 일이 생기게됩니다.

사설 이지만 Loose Coupling(느슨한 결합)을 위해서 상속도 지양해야 한다고 생각합니다.
상속은 부모의 메세지목록(인터페이스)를 물려 받을 수 있는 강력한 기능이지만 부모가 구현하고 있는 
메소드를 오버라이드를 할 수 있어 캡슐화에 기스가 날 수 있습니다.
부모가 구현하고 있는 메소드를 오버라이드 하게될때 부모가 자식이 오버라이드 하고 있는 메서드를 바꾸게 될 경우 끔직한 일이 발생 할 수도 있다고 생각합니다.

그래서 객체는 협력하고 있는 대상을 알아야하는 되도록 합성을 지향해야한다고 생각합니다........

goodcodings committed:
feature:password validator 기능 추가

비빌번호 유효성 검사에서 연속된 숫자 검증하는 로직 추가

goodcodings committed:
refactor:user domain helper메소드 변경

user 상태를 변경하는 helper 메소드의 리턴값 추가

goodcodings commented on app/src/main/java/com/codesoom/assignment/common/validateRuleInfo/PassWordRuleInfo.java:
```diff
@@ -0,0 +1,21 @@
+package com.codesoom.assignment.common.validateRuleInfo;
+
+public enum PassWordRuleInfo {
+    ALLOW_CONTINUOUS_NUMBER(3,"회 이상 연속된 문자열을 사용할 수 없습니다"),
```
유효성 검사를 하는 경우 요구사항이 변경할 수 있다는 생각을 관가해서 정말 단순하게 설계를 했다는걸 이제야 깨달았네요....

PasswordRuleInfo enum의 경우 유효성 검사를 하기 위한 헬퍼 클래스 사용하기 위해 만들었습니다.

이 enum의 책임은 몇회 이상 연속된 문자열, 몇회 이상 중복된 문자열 사용금지에 대한 간략적인 정보와 그에 따라 에러 메세지를 유효성 검사를 실제 실행하는 구현 로직에 정보를 주기 위한 역할을 수행하기 위해 존재하는 객체입니다.

johngrib replied:
간단하게 말하자면 연속된 숫자를 금지하겠다는 선언이죠. 때로는 상황을 있는 그대로 설명해 보는 것이 좋은 훈련이 될 거에요. 단순하게 보고, 단순하게 쓰고, 단순하게 작업하는 것도 훈련이 될 수 있습니다.

goodcodings commented on app/src/main/java/com/codesoom/assignment/domain/validator/PasswordValidator.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private boolean isOverFlowContinuousNumber(String password, int allowContinuousNumber) {
+        int stringSize = password.length();
```
추후 연속된 숫자 뿐만 아니라 연속된 문자열도 요구사항에 포함될 수 있어서 ASC코드로 연산했는데... 
지금 코드를 다시보니 너무 구체화 되어있어 범용적으로 사용이 불가능하겠네요 ㅠㅠㅠㅠ

johngrib replied:
이 메소드의 코드가 길어지는군요. 이 메소드만 분리해서 별도의 클래스로 작성하는 것을 고려해 보세요. 그렇게 한다면 여기에서는 그 클래스를 불러다 쓰는 방식을 생각할 수 있겠죠. 테스트 코드 작성도 재미있을 겁니다.

goodcodings commented:
간단한 구현 같은데 일일이 왜 이렇게 짤 수 밖에 없는지 생각하고, 생각을 공유하니 구현하는데 오래 걸리네요 ㅠㅠㅠ

johngrib commented:
> **1. 먼저 아는 대로 작업하지 않은 이유**
> 
> 제가 생각한 Domain layer의 구성은 `공연, 할인 정책, 예매`등으로 구성된 **도메인 모델**과 `할인 금액 계산하기, 예매하기`등과 객체간의 협력에 필요한 **추상화 레벨이 높은 메세지 목록**이 나열된 인터페이스가 존재 해야하며
> 
> application layer의 구성은 domain layer에서 나열된 **메세지의 목록**을 어떻게 풀어 나가야할지 자세한 구현을 해야하는 layer로 잘못 이해를 하고 있었습니다.
> 
> 그래서 코딩을 할 당시에 presentation layer의 구성요소중 하나인 컨트롤러에 application layer에 class를 아닌 domain layer에 인터페이스를 의존해야한다고 생각하게 됬습니다. 그런데 DDD의 의존계층을 보게되면 presentation layer은 application layer을 의존해야하고 application layer은 domain layer을 해야 한다고 되어있고 제가 생각한대로 presentation layer에 domain layer의 인터페이스를 표기하면 의존 관계가 엉망이 된다고 생각하게 되어 많이 혼란스러워서 어떻게 구현해야할지 몰라 기존에 작성한 ProductService 형태를 따라서 코딩을 하게 됬습니다.

어떻게 구현해야 할 지 몰라서 기존 코드를 그대로 사용하신 셈인데요, 이제부터 직접 생각하신대로 구현을 바꿔 보도록 합시다. 머릿속에 걱정과 원칙이 많으신 것 같아요. 그래서 알고 있는 것이 있어도 그런 생각들을 신경쓰느라 내가 알고 있는 것을 실천하지 못할 수 있죠.

지금은 코드숨 과정 진행중이니까 마음껏 새롭게 시도하고 실패해 보세요.
경험을 쌓는 것이 중요합니다. 지금 우리가 실제로 회사에 제출하는 코드를 만들고 있는 게 아니잖아요?

한편으로는 설명을 짧고 굵게 해보는 연습도 필요할 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/UserNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package com.codesoom.assignment;
+
+public class UserNotFoundException extends RuntimeException {
+    public UserNotFoundException(Long id) {
+        super("User not found: " + id);
+    }
```
지금은 `id`만 받고 있지만 예외 메시지를 받을 수 있도록 새로운 생성자를 추가해 봅시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/user/UserService.java:
```diff
@@ -0,0 +1,39 @@
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class UserService {
```
인터페이스를 구현하도록 수정해 봅시다.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요. 주말에는 좀 바쁘셨나보군요.
